icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java

```diff
@@ -432,22 +428,16 @@ private UnitsParser(String identifier) {
             }
 
             // Add simple units
+            String[] simpleUnits = UnitsData.getSimpleUnits();
             for (int i = 0; i < simpleUnits.length; i++) {
                 trieBuilder.add(simpleUnits[i], i + UnitsData.Constants.kSimpleUnitOffset);
 
             }
 
             // TODO: Use SLOW or FAST here?
             UnitsParser.savedTrie = trieBuilder.build(StringTrieBuilder.Option.FAST);
```

Comment:
Basically what you want is a compare-and-swap.  I think you can do this safely with the old code but with an extra synchronized block:

```java
synchronized {
    if (UnitsParser.savedTrie == null) {
        UnitsParser.savedTrie = trieBuilder.build(...);
    }
}
```

See https://stackoverflow.com/questions/7420504/threading-lazy-initialization-vs-static-lazy-initialization

Reply:
I think static block works fine too.